drivers: stepper: introduce stepper_drv api to facilitate implementing motion controllers as device drivers

jilaypandya opened the pull request:
**Update 17.10.25**
---------------------

split the stepper_drv_api out of stepper_api.
model tmc5xxx drivers like mfds.

----------------------

This PR aims to facilitate implementation of stepper motion controllers as device drivers. Hence the low-level step-dir drivers now shall implement a stepper_drv api, effectively also signifying what the IC actually supports i.e enable/disable, setting/getting microstep resolution and in some cases fault detection.

- Stepper API as all of us know it, implements motion control functions
- Stepper Drv API is responsible for the configuration of stepper_drv drivers.

| Drivers           |  Stepper Driver (stepper_drv API) | Stepper Motion Control Driver ( stepper API) |
|----------------------------|:--------------------------------------:|:----------------------------------------------------------------------------------------:|
| adi,tmc22xx       | x                                   |                                                                                          |
| allegro,a4979     |  x                                   |                                                                        |
| ti,drv84xx           | x                                   |                                                                         |
| adi,tmc50xx-stepper-drv                  |  x                                       |                                                                                     |
| adi,tmc50xx-stepper     |                                         | x                                                                                      |
| adi,tmc51xx-stepper-drv                  |  x                                       |                                                                                       |
| adi,tmc51xx-stepper     |                                         | x                                                                                      |
| zephyr,gpio-step-dir-stepper |                                        | x                                                                                      |
| zephyr,h-bridge-stepper |                                         |  x                                                                                    |

Stepper Driver Subsystem shall now have two APIs

|     | stepper  driver  api (stepper_drv) | stepper motion control api  (stepper) |
|----------|-----|-----------|
|enable / disable    |  impl |       |
|microstep_resolution     |  impl |        |
|move_to/move_by/run/stop  |   |  impl  |
|set_reference_position  |   |  impl  |
|*is_moving |   |  impl  | 
|get_actual_position|   |  impl  |
|set_reference_position|  |  impl  |


**Side effects**:
1. A lot of common code is refactored out and the stepper drivers only do what the ic supports. i.e. enable/disable, set/get ustep resolution.

- This PR should effectively resolve the PR #91584, which aims to be an intermediate step.
- This PR is inspired by the presentation of the complexity of implementing a motion-controller behind an all-knowing stepper api in PR #90748. Thanks @andre-stefanov. 
- PR #90677 will have rebase and implement stepper_drv api, since it's a stepper_drv driver instead of stepper api, if this PR gets merged before

faxe1008 commented on drivers/stepper/adi_tmc/tmc22xx.c:
```diff
@@ -19,26 +19,25 @@ struct tmc22xx_config {
 
 	LOG_DBG("Disabling Stepper motor controller %s", dev->name);
-	return gpio_pin_set_dt(&config->enable_pin, 0);
+	return gpio_pin_set_dt(&config->enable_pin, 1);
```
Same here

faxe1008 commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -121,7 +120,6 @@ static int a4979_stepper_set_micro_step_res(const struct device *dev,
 	case STEPPER_MICRO_STEP_4:
 		m0_value = 0;
 		m1_value = 1;
-		break;
```
I guess this was an accident

faxe1008 commented on drivers/stepper/gpio_stepper_driver.c:
```diff
@@ -0,0 +1,234 @@
+	const struct gpio_stepper_config *config = dev->config;
+
+	LOG_DBG("Initializing %s gpio_stepper with %d pin", dev->name, NUM_CONTROL_PINS);
+	for (uint8_t n_pin = 0; n_pin < NUM_CONTROL_PINS; n_pin++) {
```
Checking readyness via gpio_ready_dt

jilaypandya replied:
done

faxe1008 reviewed (commented):
Small little pass, will do some in depth review later on :^)

dipakgmx commented on drivers/stepper/gpio_stepper_driver.c:
```diff
@@ -0,0 +1,241 @@
+#include <zephyr/sys/__assert.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
```
Nit: `gpio_stepper_motor_controller`, just a bit too long :)

andre-stefanov commented on drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	if (data->direction == STEPPER_DIRECTION_POSITIVE) {
+		data->actual_position++;
+	} else {
+		data->actual_position--;
+	}
```
since we have only 2 enum values for direction (positive and negative), call of this function does not take into account if the stepper is moving at all or not. Calling it without a check would result in wrong position.

I would rather merge this function with `update_remaining_steps` or even execute the logic inline inside of both task functions because the actual logic is not really complex but extracting helper functions could introduce few ticks of execution overhead (function call overhead and data dereferencing).

andre-stefanov commented on drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+
+	data->dev = dev;
+
+	if (config->timing_source->init) {
```
we should return an error if `config->timing_source->init` evaluates to `false` because in this case the controller wont be usable (and we have the worker timing source as fallback for that reason).

andre-stefanov replied:
resolved

andre-stefanov commented on dts/bindings/stepper/adi/adi,tmc50xx.yaml:
```diff
@@ -85,7 +85,7 @@ properties:
 child-binding:
   include:
-    - name: stepper-controller.yaml
+    - name: stepper.yaml
```
i think only `en-gpios` and `invert-direction` should be allowed for this driver. Looking into the specs i dont see other features available over gpio (microstep config, step/dir etc)

andre-stefanov replied:
oh ok. probably the model i checked was an exception and did not have those pins exposed

andre-stefanov replied:
rsolved

andre-stefanov commented on include/zephyr/drivers/stepper.h:
```diff
@@ -272,6 +302,56 @@ static inline int z_impl_stepper_disable(const struct device *dev)
+ * @details This function will cause the stepper to do one micro-step.
+ * This function is typically used in step-dir stepper drivers where
+ * an external stepper motion controller is used to control the stepper.
+ * Counting the number of steps is the responsibility of the stepper motion controller,
+ * which will call this function for each step.
```
```suggestion
 * This function will not increment/decrement any position related data. Doing so
 * is responsibility of the caller (e.g. stepper motion controller).
```

jilaypandya replied:
your text sounds better :)

jilaypandya replied:
done

andre-stefanov replied:
resolved

andre-stefanov commented on include/zephyr/drivers/stepper.h:
```diff
@@ -326,6 +406,33 @@ static inline int z_impl_stepper_get_micro_step_res(const struct device *dev,
+ * @retval -ENOSYS If not implemented by device driver
+ * @retval 0 Success
+ */
+__syscall int stepper_set_fault_callback(const struct device *dev,
```
how is this different from `set_event_callback` ? also i don't see this API being implemented or used anywhere.

andre-stefanov commented on samples/drivers/stepper/generic/boards/nucleo_g071rb.overlay:
```diff
@@ -14,4 +22,12 @@
 			<&gpioa 7 GPIO_ACTIVE_HIGH>;	/* D11 */
 	};
+
+	stepper_motion_control: stepper_motion_control {
```
```suggestion
	gpio_stepper_motion_control: stepper_motion_control {
```

jilaypandya replied:
because it toggles step pin?

andre-stefanov replied:
it's just because it is a motion controller for the gpio stepper. if you have multiple steppers in a project, you will have to distinguish them. so i proposed to name them after the stepper they control.

jilaypandya replied:
I will be naming it as the driver itself `zephyr_stepper_motion_control`.

andre-stefanov commented on samples/drivers/stepper/generic/boards/nucleo_g071rb.overlay:
```diff
@@ -1,11 +1,19 @@
 / {
 	aliases {
 		stepper = &gpio_stepper;
+		stepper-motion-control = &stepper_motion_control;
```
```suggestion
		stepper-motion-control = &gpio_stepper_motion_control;
```

jilaypandya commented on dts/bindings/stepper/stepper.yaml:
that would be en-pin, step-pin, dir-pin, m0/m1/m2-gpios and maybe fault pin as well. Hmm sounds fair. Is it okay if we do this in an upcoming patch, since I am going to separate the relevant functions in stepper-step-dir-api :)?

jilaypandya replied:
we might need a patch in between for tmc22xx as well, since it handles the mix-gpios in a slight different pattern, but we can do it.

andre-stefanov commented on drivers/stepper/adi_tmc/tmc22xx.c:
```diff
@@ -19,26 +19,25 @@ struct tmc22xx_config {
 
 	LOG_DBG("Enabling Stepper motor controller %s", dev->name);
-	return gpio_pin_set_dt(&config->enable_pin, 1);
+	return gpio_pin_set_dt(&config->enable_pin, 0);
```
yes this is exactly what i came up with in my implementation in the end. It is a bit inconvenient because the user of e.g. tmc2209 has to know this inverted EN logic during DTS definition instead of just saying "i define the pins, the driver knows which level has to be used for enable/disable". But unfortunately i could not provide gpio port/pin references without output level flag in DTS so i left it as faxe is proposing.

andre-stefanov commented on drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+	struct stepper_motion_control_data *data = dev->data;
+
+	update_remaining_steps(dev);
+	stepper_step(config->stepper);
```
resolved :)

jilaypandya commented:
Hi Everyone, I have introduced `step_dir_driver_api` effectively creating two families of stepper drivers, as following:

| Drivers           |  Stepper Step-Dir Driver (step-dir API) | Stepper Motion Control Driver (Integrated Motion Control + Stepper Driver / stepper API) |
|----------------------------|:--------------------------------------:|:----------------------------------------------------------------------------------------:|
| adi,tmc22xx       | x                                   |                                                                                            |
| allegro,a4979     |  x                                   |                                                                        |
| ti,drv84xx           | x                                   |                                                                         |
|zephyr,gpio-stepper(h-bridge based driver)               | x                                        |                                                                                      |
|tmc50xx     |                                        | x                                                                                      |
| tmc51xx                  |                                        | x                                                                                      |
| zephyr,stepper-motion-controller (CPU Based)   |                                        | x                                                                                      |

jbehrensnx reviewed (commented):
Mostly minor stuff from me. Overall I am pretty happy with it and I think for an interim solution it is fine. 

There are some aspects of the final intended architecture that I feel are worth discussing, but I also feel that they would be beyond the scope of this PR.
The one exception to this is the fault event, as I feel that motion-controllers should be able to raise fault events as well.

Personally, I would prefer if the `stepper_motion_controller` got renamed to indicate that it is for step-dir drivers, but that is not a hard requirement for me.

jilaypandya commented on tests/drivers/stepper/drv84xx/api/src/main.c:
```diff
@@ -87,66 +87,6 @@ ZTEST_F(drv84xx_api, test_actual_position_set)
-	zassert_equal(pos, 50u, "Target position should be %d but is %d", 50u, pos);
-}
-
-ZTEST_F(drv84xx_api, test_move_to_negative_direction_movement)
```
so the thing is move_to wraps move_by, and for move_by we do check in both the directions, that's why I dropped this test, I remember from one of the earlier PRs that you had recommended to drop these tests

jilaypandya replied:
If you go through all the drivers, move_by is either getting wrapped by move_to or move_to is wrapping move_by. I agree that the wrap is not on the api level, but I really feel we need to have a nice balance between development and testing.

jbehrensnx replied:
The thing is, as long as something is not a wrap or check on api level, a test is warranted, as the check/wrap is not enforced by the architecture. That is the sticking point for me. I have no issue with such a check/wrap on api level and I feel like we should have probably done that this way when we first talked about wrapping move_to and move_by.

jilaypandya replied:
You can always create a patch :) If you feel something is not correct on the API level :)

Not sure how other collaborators feel about it, but as I have mentioned before, putting so many redundant test cases slows down the development/refactoring of the api especially during the experimental stage.

would it be okay for now if you handle transferring tests in the test generalization PR you are working on, I'd like to focus on getting to the further steps in the RFC, that is not going to be an easy task as well :)

jilaypandya commented on tests/drivers/stepper/drv84xx/api/src/main.c:
```diff
@@ -87,66 +87,6 @@ ZTEST_F(drv84xx_api, test_actual_position_set)
-	zassert_equal(pos, -50, "Target position should be %d but is %d", -50, pos);
-}
-
-ZTEST_F(drv84xx_api, test_move_to_identical_current_and_target_position)
```
So I have added a check in move_by, if move_by(0) is called then the function returns 0 immediately. I mean there is no reason to call move_by(0) anyhow since we have a stop function

jilaypandya commented on dts/bindings/stepper/zephyr,stepper-motion-control.yaml:
```diff
@@ -0,0 +1,33 @@
+compatible: "zephyr,stepper-motion-control"
+
+properties:
+  step-tick-ns:
```
Fair enough. I'll drop it.

jbehrensnx reviewed (commented):
Looks good so far. I currently do not have the time for a code deep dive, so I will wait for others for their opinion on this.

One thing to note is the name step_dir for the sub api. It is fine for now, as all current stepper drivers fall into that category, but for the end rework, I would change it to something like stepper_driver, as we will want to support more than step-dir drivers for stepper drivers. The other rework PRs already did this, and if you are already aware of this, sorry to bring it up again.

jilaypandya commented:
> One thing to note is the name step_dir for the sub api. It is fine for now, as all current stepper drivers fall into that category, but for the end rework, I would change it to something like stepper_driver, as we will want to support more than step-dir drivers for stepper drivers.

That's a good suggestion thanks, I have renamed step-dir api to stepper_drv api to better reflect its intended use.

jilaypandya commented on drivers/stepper/adi_tmc/tmc50xx.c:
```diff
@@ -25,6 +25,7 @@ struct tmc50xx_config {
 	const uint32_t gconf;
 	struct spi_dt_spec spi;
 	const uint32_t clock_frequency;
+	const struct device **steppers;
```
stallguard_dwork and rampstat_callback_dwork are the part of stepper_data, and they need be accessed while executing move_to and run, that's why the controller needs stepper instances.

jilaypandya commented on drivers/stepper/adi_tmc/tmc51xx/tmc51xx.h:
```diff
@@ -44,12 +37,29 @@ struct tmc51xx_config {
 	const struct device *stepper;
 	stepper_event_callback_t callback;
 	void *event_cb_user_data;
+	struct gpio_callback diag0_cb;
```
yes you are right i'll move it to controller

jilaypandya replied:
done

dipakgmx reviewed (commented):
A few comments from first glance. Else, it seems good

jilaypandya commented on drivers/stepper/Kconfig.stepper_event_template:
```diff
@@ -1,18 +1,18 @@
 # SPDX-License-Identifier: Apache-2.0
 
-config STEPPER_$(module)_GENERATE_ISR_SAFE_EVENTS
+config $(module)_GENERATE_ISR_SAFE_EVENTS
```
Sorry, just thought about it again, this event mechanism is getting used in zephyr,stepper-motion-controller and is independent of the stepper-drv, since motion-controller will know when the steps are completed and will trigger the events.

jilaypandya replied:
I think you meant that a tempate is not rquired anymore? I've moved it to the Kconfig.zephyr_stepper_motion_controller.

jilaypandya commented on include/zephyr/drivers/stepper.h:
```diff
@@ -482,67 +376,332 @@ static inline int z_impl_stepper_move_to(const struct device *dev, const int32_t
+ */
+
+/**
+ * @cond INTERNAL_HIDDEN
```
only the typedefs and driver struct is hidden, the stepper_drv api is visible to the user via doxygen documentation

jilaypandya commented on drivers/stepper/stepper_shell.c:
```diff
@@ -109,6 +131,34 @@ static void cmd_stepper_direction(size_t idx, struct shell_static_entry *entry)
 
 SHELL_DYNAMIC_CMD_CREATE(dsub_stepper_direction, cmd_stepper_direction);
 
+static void cmd_stepper_axis_dir(size_t idx, struct shell_static_entry *entry)
```
done

rruuaanng commented on dts/bindings/stepper/zephyr,fake-stepper.yaml:
```diff
@@ -2,9 +2,8 @@
 description: |
-    This binding provides a fake stepper controller for use as either a stub or a mock in
-    Zephyr testing.
+    This binding provides a fake stepper for use as either a stub or a mock in Zephyr testing.
```
Please remove `This binding`. When it appears in `dts/bindings`, we know it's purpose :)

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	if (!k_is_in_isr()) {
+		data->callback(dev, MAX_SUPPORTED_STEPPER, event, data->event_cb_user_data);
```
Won't that correct with 0 < MAX_SUPPORTED_STEPPER.

Would have to most probably drop the last commit for now.

jilaypandya replied:
fixed

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->actual_position = position;
```
done

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+	K_SPINLOCK(&data->lock) {
+		data->callback = cb;
+	}
+	data->event_cb_user_data = user_data;
```
fixed it.

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,405 @@
+	/* Run the callback */
+	if (data->callback != NULL) {
+		/* using MAX_SUPPORTED_STEPPER as currently only one stepper is supported */
+		data->callback(data->dev, MAX_SUPPORTED_STEPPER, event,
```
fixed

jilaypandya committed:
drivers: stepper: split stepper api

- split stepper api into stepper and stepper_drv api
- stepper api now comprises only of motion control apis
- stepper_drv api comprises of apis for configuring stepper
  drivers

- add documentation about stepper and stepper_drv api
- move stepper.rst in a dedicated stepper folder
- add information about stepper_drv api and relevant functions
  in stepper documentation.

- drop motion control functions from all the stepper_drv drivers
- create a common a library for controlling stepper motors by
  toggling gpios via h-bridge or step-dir stepper_drivers

- tmc5xxx devices are a combination of motion controller and
  stepper driver devices. tmc5xxx devices need to be modelled as
  mfds in order to address the split in stepper driver subsystem

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented:
Ping @dipakgmx, @faxe1008, @jbehrensnx could you test the code on your hw setup as well :) ? Thanks :)

dipakgmx commented on drivers/stepper/adi_tmc/tmc50xx.c:
```diff
@@ -93,22 +96,33 @@ static int tmc50xx_read(const struct device *dev, const uint8_t reg_addr, uint32
+	__ASSERT_NO_MSG(controller != NULL);
+	__maybe_unused const struct tmc50xx_config *tmc50xx_config = controller->config;
+
+	CHECK_STEPPER_IDX(controller, stepper_index, tmc50xx_config->num_steppers);
```
What advantage does this have over the `__ASSERT` macro?

jilaypandya commented on drivers/stepper/adi_tmc/tmc50xx.c:
```diff
@@ -93,22 +96,33 @@ static int tmc50xx_read(const struct device *dev, const uint8_t reg_addr, uint32
 
-static int tmc50xx_stepper_set_event_callback(const struct device *dev,
-					      stepper_event_callback_t callback, void *user_data)
+static void check_stepper_idx(const struct device *controller, const uint8_t stepper_index)
```
There are quite a few functions where i have to access the `controller_config` just for the number of steppers, hence i wrapped up the macro like this.

jbehrensnx reviewed (changes requested):
First of all, it worked on my end. I tested it with the drv84xx.
The reworked architecture also looks fine to me.

Most of the things I commented on are effectively one-liners or are non-blocking.

The big thing is the folder structure of the drivers. As mentioned in the comment, having two drivers implementing different apis next to each other in the same folder just doesn't sit right with me and we might want to reorganize the folder structure.

But overall I am quite happy with how this turned out.

jilaypandya commented on doc/hardware/peripherals/stepper.rst:
```diff
@@ -3,22 +3,26 @@
 ########
 
-The stepper driver API provides a set of functions for controlling and configuring stepper drivers.
+The stepper driver subsystem consists of two APIs:
```
In the context of the Zephyr OS, a [subsystem](https://docs.zephyrproject.org/latest/glossary.html#term-subsystem) refers to a logically distinct part of the operating system that handles specific functionality or provides certain services. Subsystems can include components such as networking, file systems, **_device driver classes_,** power management, and communication protocols, among others. Each subsystem is designed to be modular and can be configured, customized, and extended to meet the requirements of different embedded applications.

https://docs.zephyrproject.org/latest/introduction/index.html

jilaypandya commented on tests/drivers/build_all/stepper/app.overlay:
```diff
@@ -39,5 +39,17 @@
+		zephyr_stepper_motion_control_counter: zephyr_stepper_motion_control_counter {
+			compatible = "zephyr,stepper-motion-control";
+			status = "okay";
+			stepper = <&zephyr_h_bridge_stepper>;
```
done

jilaypandya commented on dts/bindings/stepper/zephyr,stepper-motion-control.yaml:
```diff
@@ -0,0 +1,26 @@
+properties:
+  counter:
+    type: phandle
+    description: Counter used for generating step-accurate pulse signals.
```
done

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,404 @@
+	}
+
+	if (!k_is_in_isr()) {
+		data->callback(dev, MAX_SUPPORTED_STEPPER - 1, event, data->event_cb_user_data);
```
done

jilaypandya commented on drivers/stepper/zephyr_stepper_motion_controller.c:
```diff
@@ -0,0 +1,404 @@
+	}
+}
+
+static int z_stepper_motion_control_move_by(const struct device *dev, const uint8_t stepper_idx,
```
Yeah, that's a short for zephyr_stepper_motion_control, the DT_DRV_COMPAT for this driver.

jilaypandya commented on drivers/stepper/stepper_shell.c:
```diff
@@ -499,19 +651,21 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
+		      cmd_stepper_run, 4, 0),
+	SHELL_CMD_ARG(stop, &dsub_pos_stepper_controller_name, "<device>", cmd_stepper_stop, 3, 0),
+	SHELL_CMD_ARG(control_info, &dsub_pos_stepper_controller_name, "<device>",
+		      cmd_stepper_control_info, 3, 0),
```
Don't have a strong opinion on this. Yeah it can be split into stepper_shell and stepper_drv_shell.

faxe1008 commented:
Checked this on the TMC2209, looks good from my side. Only little nit I see overall is the naming, not a fan of abbreviations `stepper_drv_*`. In this day and age we are not limited by any max symbol name length anymore :^)

jbehrensnx commented:
> If i go for stepper_driver_* then the current stepper_drv_driver_api would be stepper_driver_driver_api, to me it sounds a bit funny :). We can for sure think of another name :)

It sounds funny, but it would be correct. The physical drv8424 is, for example, a stepper driver, so the drv84xx driver is a stepper driver driver, even if it sounds strange and I am also sometimes uncertain to use it in discussions because of this.

Something that sometimes annoys me a bit is the fuzzy using of terms. We have on the physical side:

- `stepper motors`
- `stepper drivers` (tmc2209, drv8424, ...)
- `stepper (motion) controllers` (tmc50xx series, tmc429, ...)

And on the software side

- `stepper driver drivers` (tmc22xx, drv84xx, ...), with the h-bridge driver possibly being a `stepper driver` instead
- `stepper (motion) controllers` (zephyr_stepper_motion_controller)
- `stepper (motion) controller drivers` (tmc50xx, tmc51xx, ...)

And I feel like we should all make an effort to use the correct terms, even if they sound a bit strange. There is for example no "`stepper`", it could be any of the mentioned components and thus should be clarified.

That said, parts of this is out of scope i.m.o. for this PR and might be better served with a separate issue/PR.
Personally I would be fine with a stepper driver driver api, it sounds strange but is the correct term.
There is also the unresolved comment from my review concerning the documentation comments. But these two are i.m.o. the only two things worth discussing for this PR.

The rest should be, i.m.o.,  discussed outside this PR.

faxe1008 commented:
> If i go for `stepper_driver_*` then the current `stepper_drv_driver_api` would be `stepper_driver_driver_api`, to me it sounds a bit funny :). We can for sure think of another name :)

True, have not thought about that. Fine with `drv` then :^)

jilaypandya commented:
> True, have not thought about that. Fine with `drv` then :^)

Thx :).

jbehrensnx commented on include/zephyr/drivers/stepper.h:
```diff
@@ -234,170 +218,87 @@ __subsystem struct stepper_driver_api {
  * @brief Set the reference position of the stepper
  *
  * @param dev Pointer to the stepper driver instance.
+ * @param stepper_index The index of the stepper to set the reference position for.
```
In that case, use stepper motor instead. Just stepper by itself is not enough, after all that could mean stepper motor or stepper driver. That is my main issue with the second line.

jbehrensnx reviewed (approved):
Fine for me

jbehrensnx commented on tests/drivers/stepper/drv84xx/api/src/main.c:
```diff
@@ -240,13 +209,12 @@ ZTEST_F(drv84xx_api, test_run_positive_direction_correct_position)
-	(void)stepper_get_actual_position(fixture->dev, &steps);
-	zassert_true(IN_RANGE(steps, 4, 6), "Current position should be between 4 and 6 but is %d",
+	(void)stepper_get_actual_position(fixture->dev, CONFIG_STEPPER_IDX, &steps);
+	zassert_true(IN_RANGE(steps, 1, 6), "Current position should be between 4 and 6 but is %d",
```
```suggestion
	zassert_true(IN_RANGE(steps, 1, 6), "Current position should be between 1 and 6 but is %d",
```

Missed this previously (same goes for negative direction as well).
Also, why did you go from 4 to 1?

jbehrensnx replied:
Just reran these tests on native sim a couple of times on my pc (including random ordering) and they all passed, even with the 4-6 range.

Concerning the accuracy, that is because these tests were originally developed for tests on real hardware.

jilaypandya replied:
Okay then let's revert this change :) Probably was a mixup with QEMU.

jbehrensnx reviewed (changes requested):
Missed a wrong assert message in the drv84xx/api tests in my last reviews.

dipakgmx commented on tests/drivers/stepper/drv84xx/api/src/main.c:
```diff
@@ -240,12 +209,11 @@ ZTEST_F(drv84xx_api, test_run_positive_direction_correct_position)
-	(void)stepper_run(fixture->dev, STEPPER_DIRECTION_POSITIVE);
+	(void)stepper_set_microstep_interval(fixture->dev, CONFIG_STEPPER_IDX, step_interval);
+	(void)stepper_run(fixture->dev, CONFIG_STEPPER_IDX, STEPPER_DIRECTION_POSITIVE);
 	k_busy_wait(110000);
```
Not related to the changes introduced by this PR, but why is a `k_busy_wait` used here? Are we sure, we don't want to relinquish control of a test function for such a long time?

jbehrensnx replied:
That was an error made by me a while ago. I think it was in an attempt to have an accurate wait time (which is somewhat relevant for this test). That said, it causes issues with work queue timing and I am fixing it in the test generalization. Updating said PR for these changes is going to be fun.

faxe1008 commented on doc/hardware/peripherals/stepper.rst:
```diff
@@ -3,22 +3,26 @@
 
-The stepper driver API provides a set of functions for controlling and configuring stepper drivers.
+The stepper driver subsystem consists of two device driver APIs:
 
```
Maybe an explanation which driver is responsible for the `step` function?

And an explanation that the stepper driver API makes use of the steper_drv API internally? Some sort of diagram maybe with a driver which implements both vs one where you have to use `zephyr,stepper-motion-control`

dipakgmx replied:
+1 for some diagrams

jilaypandya replied:
I will draw up something nice :)

jilaypandya commented:
Hi Everyone, I  have introduced one more commit. I think we should bring it in this PR, so that we don't have to introduce breaking changes in the PR again and again. Sorry for extending the scope of the PR again, but I think it's better that we bring this change in this PR. The most affected drivers are the trinamic ones for now :)
```
drivers: stepper: introduce event cb mechanism for stepper_drv
stall detection is a stepper_drv functionality and hence stepper_drv
api needs an event callback mechanism
```

jbehrensnx reviewed (commented):
I have no time for a full review at this point, but at a fist glance I see no issues, through I do have a question (see below).

jilaypandya commented on include/zephyr/drivers/stepper.h:
```diff
@@ -105,14 +105,12 @@ enum stepper_run_mode {
 	/** Steps set using move_by or move_to have been executed */
 	STEPPER_EVENT_STEPS_COMPLETED = 0,
-	/** Stall detected */
-	STEPPER_EVENT_STALL_DETECTED = 1,
```
Stall detection seems to be a part of stepper drivers and not stepper motion controllers, see in the image below, depending on a certain configuration of tmc5130 driver, the motion control part gets disabled however stall detection since its part of the stepper driver block is still present. 

<img width="435"  alt="image" src="https://github.com/user-attachments/assets/0e7e70d3-38f4-4008-8896-48352d118681" />

The same goes for tmc2130 (a stepper driver ) as well :)

<img width="435"  alt="image" src="https://github.com/user-attachments/assets/d00a4a0a-99e6-4950-b6b2-6d0a45957f08" />

jbehrensnx replied:
The question was more about whether stepper controllers should have a separate stall event as well. The stepper_drv should definitely have it.

A quick look at their datasheets show that TMC457 and 4361 both process stallguard events. And with these two controllers the stepper driver would not necessarily appear as a device in zephyr.
Thus both stepper controllers and stepper drivers having a stall event might be a good idea.

jilaypandya replied:
> The question was more about whether stepper controllers should have a separate stall event as well.

The event as of now is STALL_EVENT_DETECTED, which from what I've seen can only be detected by stepper drivers

> A quick look at their datasheets show that TMC457 and 4361 both process stallguard events.

A detailed look at the datasheets shows that these controllers process and not detect stall. It seems that all these controllers can **process** stall detection from stepper drivers and then trigger events such as stop_sg_1 or stop_on_stall.
Quoting Trinamic datasheets below :)

```
TMC457

The TMC457 has a driver SPI to directly control the TRINAMIC stepper motor 
drivers TMC236, TMC239, TMC246, and TMC249. It supports processing of StallGuard information to 
emulate a reference switch,

13 stop_SG 1 Stop motor on StallGuard event signaled by TMC246 / TMC249
```
```
TMC4361

STALL DETECTION HANDLING
- TMC motor driver chips always return the stall detection status to the TMC4361 motion controller 

29 STOP_ON_STALL: Motor stall detected. Motor has been stopped. 
```

So the processing of the stall detected by stepper drivers is done by controllers and then this info can be propogated using a generic event called `STEPPER_EVENT_STOPPED`.

On a general note, `stepper_events` should be as generic as possible, if some individual `stepper motion controller` can `detect` a stall on itself, it would have to go via custom events. 

We could probably down the line add a stepper event called `STOPPED_ON_STALL`. But adding new stepper_events is something we can do in another dedicated PR.

So answering your question :) I see `STALL_DETECTED` is a stepper_driver event and not a stepper event. Please note my emphasis on the word `detect`.  The `STALL_EVENT_DETECTION` is a functionality offered by a lot of trinamic stepper drivers and even a [ST driver](https://www.st.com/resource/en/datasheet/l9942.pdf) (Probably there are more)

jbehrensnx replied:
Fair enough.

jilaypandya commented:
Hey Everyone,

I am working towards dropping the stepper_idx from the stepper_api. 
Devices like the TMC5XXX would need to be modeled as a MFD, I have pushed the changes for TMC50XX. I will be working on the TMC51XX in the coming days. Thanks :)

jbehrensnx commented on drivers/stepper/Kconfig.fake:
```diff
@@ -4,8 +4,8 @@
 
 config FAKE_STEPPER
-	bool "Fake stepper driver"
+	bool "Fake stepper"
```
I would  use controller & driver instead, like you do in the help.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc50xx/Kconfig.tmc50xx:
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config STEPPER_ADI_TMC50XX
+	bool "Activate trinamic tmc50xx stepper driver"
```
This isn't the stepper driver, its the mfd.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc50xx/tmc50xx.c:
```diff
@@ -0,0 +1,309 @@
+	struct spi_dt_spec spi;
+	const uint32_t clock_frequency;
+	const struct device **stepper_drivers;
+	uint8_t num_steppers;
```
```suggestion
	uint8_t num_stepper_drivers;
```

jbehrensnx commented on drivers/stepper/adi_tmc/tmc50xx/tmc50xx_stepper_driver.c:
```diff
@@ -0,0 +1,184 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(tmc50xx, CONFIG_STEPPER_LOG_LEVEL);
+
+struct tmc50xx_stepper_data {
```
NIT: I would recommend using `stepper_driver` (e.g. `tmc50xx_stepper_driver_data`) throughout the file to make it clear, which tmc50xx functionality is implemented.

jbehrensnx reviewed (changes requested):
Looks good overall.
Mostly minor changes or nitpicks, except for the one relating to `step_width_ns` (see that comment).

jilaypandya commented on drivers/stepper/gpio_stepper/gpio_step_dir.c:
```diff
@@ -0,0 +1,288 @@
+		.common.timing_source_cb = stepper_handle_timing_signal,                           \
+		.step_pin = GPIO_DT_SPEC_INST_GET(inst, step_gpios),                               \
+		.dir_pin = GPIO_DT_SPEC_INST_GET(inst, dir_gpios),                                 \
+		.step_width_ns = DT_INST_PROP(inst, step_width_ns),                                \
```
> First of all, I would recommend that this variable has a default value if it is not set in any other way (e.g. 1000 ns). Each stepper driver also should have its own default value (which is the current state), as users shouldn't have to get them from the datasheets every time. And of course users should be able to set their own value.

Default values in DT require a reason. We have defaults for step-width-ns in tree, since the stepper-drivers have ingrained motion-control logic. Now once this split happens, it is the responsibility of the user to configure the `gpio-step-dir-controller` appropriately in order to generate pulses, that's also the reason, why this property has been made required now.

step-width-ns logically is part of the pulse generation logic and hence is the part of `gpio-step-dir-controller`.

jbehrensnx replied:
The thing is, motion controller and stepper driver are, on a hardware level, already tightly coupled. So there is, from my point of view, no issue on having them tightly coupled on the Zephyr level if both are defined in Zephyr. 

This is why I mentioned that the pointer property should be optional: If only the motion controller is defined in Zephyr, a coupling is obviously not possible, but if both are present, this coupling only reflects the hardware reality.

If it wouldn't create a ton of other issues, I would recommend making the stepper driver a child of the motion controller.

Concerning the `step-width-ns` property: Having a speed limit attached to the stepper driver makes sense, as it has a physical limit, either by its very nature (e.g. 2ns for the drv84xx) or because of the properties of the stepper motor (which is currently not modeled). Yes, the `step-width-ns` property is used by the motion controller, but it is partly a property of the device the motion controller  controls, which is the stepper driver.

jilaypandya commented:
rebased on main and addressed the latest comments. Once we are through with the reviews, I would squash the commits together, write the migration guide and remove the `DNM` label.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc50xx/tmc50xx.c:
```diff
@@ -655,125 +247,63 @@ static int tmc50xx_init(const struct device *dev)
+		TMC50XX_CHILD_DEVICES_ARRAY(inst, adi_tmc50xx_motion_controller);                  \
+	BUILD_ASSERT(ARRAY_SIZE(tmc50xx_motion_controllers_##inst) <= 2,                           \
+		     "tmc50xx can drive two steppers at max");                                     \
+	BUILD_ASSERT(ARRAY_SIZE(tmc50xx_stepper_drivers_##inst) <= 2,                              \
```
Unless I misunderstood how the tmc50xx works, you should also check that the number of motion controllers and stepper drivers is the same, as they are in a 1-to-1 relationship.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc51xx/tmc51xx.c:
```diff
@@ -337,55 +393,6 @@ static void __maybe_unused tmc51xx_diag0_gpio_callback_handler(const struct devi
-	return 0;
-}
-
 int tmc51xx_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
```
Rename the function, as we only have motion controllers or stepper drivers, but no steppers.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc51xx/tmc51xx.h:
```diff
@@ -7,59 +7,73 @@
 
-#include <adi_tmc_bus.h>
+/**
+ * @brief Trigger the registered callback for stepper events.
```
```suggestion
 * @brief Trigger the registered callback for stepper driver events.
```

jbehrensnx commented on drivers/stepper/adi_tmc/tmc51xx/tmc51xx_stepper_driver.c:
```diff
@@ -0,0 +1,181 @@
+	return 0;
+}
+
+static int tmc51xx_stepper_enable(const struct device *dev)
```
```suggestion
static int tmc51xx_stepper_driver_enable(const struct device *dev)
```
Also check other functions and macros in this file.

jbehrensnx commented on drivers/stepper/gpio_stepper/common/include/gpio_stepper_common.h:
```diff
@@ -0,0 +1,290 @@
+#include "stepper_timing_source.h"
+
+/**
+ * @brief Common GPIO stepper config.
```
```suggestion
 * @brief Common GPIO motion controller config.
```
Think about renaming everything related to this to GPIO motion controller, which is admittedly quite a lot of stuff.
Everything here is about motion controllers, so I feel the names should reflect this.

jbehrensnx reviewed (changes requested):
Lots of comments, but they are primarily about wording in comments and function names. The main thing is clarifying if something is about a stpper driver or a motion controller.

On a functional level the big things are the fact that we need a new `stepper_drv_api` testsuite if the stepper_drv specific tests are removed from the `stepper_api` testsuite, and the step_width limits from devicetree, for which I will post some additional thoughts shortly in the previous discussion.

Despite all of my critique, I am overall quite happy with the state of the rework.

jilaypandya commented on drivers/stepper/adi_tmc/tmc50xx/tmc50xx_motion_controller.c:
```diff
@@ -0,0 +1,448 @@
+	return config->index;
+}
+
+int tmc50xx_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
```
I see the confusion that's happening over here. I think the best would be to use the `<device>_<api>_<func>` format for such devices. For instance, tmc51xx_stepper_move_to and tmc51xx_stepper_drv_enable.

jbehrensnx replied:
I would use motion_controller instead of stepper, so `tmc51xx_motion_controller_move_to`. But yes to the overall idea.

Edit: The reason why I would use motion_controller instead of stepper is the fact, that they are of the device class motion controller. Ideally we would have the two apis of stepper_drv and stepper_motion_controller, but I keeping the old api name for one of them is the better decision in my opinion.

jbehrensnx reviewed (changes requested):
Nearly completely only things that fell through the cracks during renaming or documentation that wasn't updated with the latest changes.

Somewhat surprised, that the new `stepper-drv` wasn't optional, but I can see that this simplifies the code for the relevant properties.
That said, I left some comments in regard of possibly slightly expanding `fake-stepper-drv` to allow it to fill the role for a stepper-drv stub in that regard. This would also allow for a basic level of usage of stepper-drv who don't yet have a  Zephyr driver. `fake-stepper-drv` should probably be renamed in that case however to clarify that it is a stub/stand-in.

jbehrensnx commented on drivers/stepper/adi_tmc/tmc51xx/tmc51xx_stepper_drv.c:
```diff
@@ -0,0 +1,181 @@
+	data->drv_event_cb(dev, event, data->drv_event_cb_user_data);
+}
+
+static int tmc51xx_set_stepper_drv_event_callback(const struct device *stepper,
```
```suggestion
static int tmc51xx_stepper_drv_set_event_callback(const struct device *stepper,
```

jbehrensnx commented on drivers/stepper/fake_stepper_controller.c:
```diff
@@ -117,10 +125,18 @@ static void fake_stepper_reset_rule_before(const struct ztest_unit_test *test, v
+	return 0;
+}
+
+static int fake_stepper_controller_init(const struct device *dev)
```
```suggestion
static int fake_stepper_init(const struct device *dev)
```

bjarki-andreasen reviewed (dismissed):
Design looks good, makes sense from a devicetree model perspective to split the device model like this

jbehrensnx commented on dts/bindings/stepper/zephyr,gpio-step-dir-stepper.yaml:
```diff
@@ -0,0 +1,53 @@
+    type: phandle
+    description: |
+      Reference to the stepper driver device.
+      This property is required in order to get the minimum step pulse width and
```
NIT: Now that this property is optional, a slight rewording might be a good idea.

jbehrensnx reviewed (changes requested):
Only minor things remaining, functional it LGTM.

jilaypandya committed:
doc: migration_guide: address stepper split

add entry for splitting stepper api in migration-guide-4.4

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jbehrensnx commented on include/zephyr/drivers/stepper.h:
```diff
@@ -235,105 +162,10 @@ __subsystem struct stepper_driver_api {
  * @brief Set the reference position of the stepper
  *
- * @param dev Pointer to the stepper driver instance.
+ * @param dev Pointer to the stepper controller instance.
```
```suggestion
 * @param dev Pointer to the stepper instance.
```
Not sure what is the best decision in this (and the other similar cases), but this way it is consistent with the rest of the changes

jbehrensnx replied:
I don't consider this resolved with the latest changes, because while `Pointer to the stepper driver instance.` is technically correct, it causes confusion with the stepper_drv api, which is for stepper drivers. One option would be to reword the sentence, while the other one would be to use a sentence based on the other apis, e.g. `Pointer to the device structure for the driver instance.` or `Stepper device instance.`

Otherwise, everything else looks good to me.

jilaypandya commented:
- addressed recent naming comment
- refined migration guide

faxe1008 reviewed (dismissed):
I think 239 comments on a PR might be a new record :^). It looks like this is finally ready. Congrats on finishing this and for your endurance throughout the process.
